File: src/patterns/stringCompile.ts
import { makeRe } from "minimatch"

import type { PatternMinimatch, Pattern } from "./pattern.js"

/**
 * @since 0.8.0
 */
export type StringCompileOptions = {
	/**
	 * Disables case sensitivity.
	 *
	 * @default false
	 *
	 * @since 0.8.0
	 */
	nocase?: boolean
}

/**
 * Compiles a string of the {@link Pattern}.
 *
 * @see {@link patternCompile}
 *
 * @since 0.8.0
 */
export function stringCompile(
	pattern: string,
	context: Pattern = [],
	options?: StringCompileOptions,
): PatternMinimatch {
	const original = pattern
	if (pattern.endsWith("/")) {
		pattern = pattern.substring(0, pattern.length - 1)
	}
	if (pattern.startsWith("/")) {
		pattern = pattern.substring(1)
	} else if (!pattern.startsWith("**/")) {
		pattern = "**/" + pattern
	}

	pattern += "/**"

	const re = makeRe(pattern, {
		dot: true,
		nonegate: true,
		nocomment: true,
		nobrace: true,
		nocase: options?.nocase ?? false,
	}) as RegExp

	return { re, pattern: original, patternContext: context }
}

File: internal/shared/stringCompile.go
package shared

import (
	"strings"

	"github.com/bmatcuk/doublestar/v4"
)

type StringCompileOptions struct {
	// Disables case sensitivity.
	//
	// Default:
	//  false
	//
	// # Since 0.8.0
	NoCase bool
}

// Compiles a string of the {@link Pattern}.
//
// See [Pattern.Compile].
//
// # Since 0.8.0
func StringCompile(
	pattern string,
	context Pattern,
	options StringCompileOptions,
) PatternMinimatch {
	original := pattern
	if strings.HasSuffix(pattern, "/") {
		pattern += "**"
	}
	if strings.HasPrefix(pattern, "/") {
		pattern = pattern[1:]
	} else if !strings.HasPrefix(pattern, "**/") {
		pattern = "**/" + pattern
	}
	if !strings.HasSuffix(pattern, "/**") {
		pattern = pattern + "/**"
	}

	var re func(path string) (bool, error)
	if options.NoCase {
		pattern = strings.ToLower(pattern)
		re = func(path string) (bool, error) {
			path = strings.ToLower(path)
			return doublestar.Match(pattern, path)
		}
	} else {
		re = func(path string) (bool, error) {
			return doublestar.Match(pattern, path)
		}
	}
	return PatternMinimatch{
		Re:             re,
		Pattern:        original,
		PatternContext: context,
	}
}
